File: services/collector/src/wb/msk-day.ts
const MSK_UTC_OFFSET_MINUTES = 180;
const ZONELESS_DATETIME = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?$/;
const ZONELESS_DATE = /^\d{4}-\d{2}-\d{2}$/;

/**
 * The single place that converts an instant into a Europe/Moscow calendar day
 * (AD-7). WB `date`/`lastChangeDate` arrive as Moscow wall-clock text without a
 * zone, so zoneless input is read as Moscow time, never as the runtime zone.
 */
export function mskDay(value: string | Date, now: Date = new Date()): string {
  if (value instanceof Date) return moscowDate(value.getTime(), now);
  if (typeof value !== 'string') throw new RangeError(`mskDay: unsupported input ${JSON.stringify(value)}`);
  if (ZONELESS_DATE.test(value)) return value;
  if (ZONELESS_DATETIME.test(value)) return value.slice(0, 10);
  const instant = Date.parse(value);
  if (Number.isNaN(instant)) throw new RangeError(`mskDay: invalid instant ${JSON.stringify(value)}`);
  return moscowDate(instant, now);
}

export function mskToday(now: Date = new Date()): string {
  return moscowDate(now.getTime(), now);
}

/**
 * Zoneless WB wall-clock text (`2026-08-17T06:48:49`) as an unambiguous
 * instant: the same text with the fixed Moscow offset. Moscow has had no DST
 * since 2014, so the mapping is bijective and a stored `last_change_at` can be
 * turned back into a WB `dateFrom` cursor (Story 1.5 --resume). Input that
 * already carries a zone or has another shape is rejected, never guessed.
 */
export function mskInstant(value: string): string {
  if (typeof value !== 'string' || !ZONELESS_DATETIME.test(value)) {
    throw new RangeError(`mskInstant: expected zoneless Moscow datetime, got ${JSON.stringify(value)}`);
  }
  return `${value}+03:00`;
}

function moscowDate(epochMilliseconds: number, now: Date): string {
  return new Date(epochMilliseconds + moscowOffsetMinutes(now) * 60_000).toISOString().slice(0, 10);
}

/**
 * Europe/Moscow offset in minutes east of UTC, taken from the runtime zone
 * database on purpose: if Moscow ever reintroduces DST, only this helper has
 * to notice.
 */
function moscowOffsetMinutes(now: Date): number {
  let parts: Intl.DateTimeFormatPart[];
  try {
    parts = new Intl.DateTimeFormat('en-US', {
      timeZone: 'Europe/Moscow',
      year: 'numeric', month: '2-digit', day: '2-digit',
      hour: '2-digit', minute: '2-digit', second: '2-digit', hourCycle: 'h23',
    }).formatToParts(now);
  } catch {
    return MSK_UTC_OFFSET_MINUTES;
  }
  const field = (type: Intl.DateTimeFormatPartTypes): number =>
    Number(parts.find((part) => part.type === type)?.value ?? '');
  const year = field('year');
  const month = field('month');
  const day = field('day');
  // An engine that ignores `hourCycle` and renders midnight as hour 24 keeps
  // the calendar day of the instant, so only the hour is folded back to 0.
  // Nothing else may be rewritten: a blanket `'24' -> '00'` on every part
  // also zeroed minute :24 and the 24th day of the month, shifting the
  // sampled offset by 24 minutes or a full day (KF-3).
  const hour = field('hour') % 24;
  const minute = field('minute');
  const second = field('second');
  if ([year, month, day, hour, minute, second].some((value) => !Number.isSafeInteger(value))) {
    return MSK_UTC_OFFSET_MINUTES;
  }
  const wallAsUtc = Date.UTC(year, month - 1, day, hour, minute, second);
  const truncatedNow = Math.floor(now.getTime() / 1000) * 1000;
  return Math.round((wallAsUtc - truncatedNow) / 60_000);
}

File: services/control-plane/src/proxima_control_plane/common/msk_day.py
"""The single place that converts an instant into a Europe/Moscow calendar day.

Python twin of the collector helper ``src/wb/msk-day.ts`` (AD-7): both must agree
on every input, otherwise the norm job and the collector fold the same WB answer
into different days. WB ``date``/``lastChangeDate`` arrive as Moscow wall-clock
text without a zone, so zoneless input is read as Moscow time, never as the
runtime zone.
"""

from __future__ import annotations

import re
from datetime import date, datetime, timedelta, timezone
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

_ZONELESS_DATETIME = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?$")
_ZONELESS_DATE = re.compile(r"^\d{4}-\d{2}-\d{2}$")
_FIXED_MSK_OFFSET = timezone(timedelta(hours=3))

try:
    _MSK = ZoneInfo("Europe/Moscow")
except ZoneInfoNotFoundError:  # pragma: no cover - bare container without tzdata
    _MSK = _FIXED_MSK_OFFSET


def msk_day(value: str | datetime | date) -> str:
    """Return the Moscow calendar day (``YYYY-MM-DD``) of ``value``.

    Zoneless strings and naive datetimes are Moscow wall clock; zoneless dates
    are already a Moscow day. Invalid input raises ValueError instead of guessing.
    """
    if isinstance(value, datetime):
        if value.tzinfo is None:
            return value.date().isoformat()
        return value.astimezone(_MSK).date().isoformat()
    if isinstance(value, date):
        return value.isoformat()
    if not isinstance(value, str):
        raise ValueError(f"msk_day: unsupported input {value!r}")
    if _ZONELESS_DATE.match(value):
        return value
    if _ZONELESS_DATETIME.match(value):
        return value[:10]
    try:
        parsed = datetime.fromisoformat(value)
    except ValueError as exc:
        raise ValueError(f"msk_day: invalid instant {value!r}") from exc
    if parsed.tzinfo is None:
        return parsed.date().isoformat()
    return parsed.astimezone(_MSK).date().isoformat()


def msk_today(now: datetime | None = None) -> str:
    """Moscow calendar day of ``now`` (UTC now by default)."""
    instant = datetime.now(timezone.utc) if now is None else now
    return msk_day(instant)
